{
  try {
    IOException exception=null;
    for (    Future<?> future : this.futures) {
      try {
        future.get();
      }
 catch (      InterruptedException ie) {
        if (exception == null) {
          exception=new IOException(ie);
        }
        exception.initCause(ie);
      }
catch (      ExecutionException ee) {
        if (exception == null) {
          exception=new IOException(ee.getCause());
        }
      }
    }
    if (exception != null) {
      throw exception;
    }
  }
  finally {
    ExecutorsUtils.shutdownExecutorService(this.executor,Optional.of(LOGGER));
  }
}
